elastic/elastic-serverless-forwarder
# Logstash support - technical preview

## What does this PR do?

This PR introduces:
* First release of a new shipper, that enables ESF to send logs to a Logstash instance. It is intended to be a "Technology Preview".
* Unit tests and integration tests covering the most important basic scenarios.
* A new architecture diagram, listing Logstash as new possible output.

## Why is it important?

Refer to the master issue https://github.com/elastic/obs-infraobs-team/issues/809

## Checklist



- [ ] I have added an entry in `CHANGELOG.md`

## Author's Checklist



## How to test this PR locally



* Execute integration tests
` py.test -vv -s tests/handlers/aws/test_logstash_output.py`

## How to test this PR in AWS

* Install Logstash on an EC2 instance. See the official documentation https://www.elastic.co/guide/en/logstash/current/installing-logstash.html 
* Install [ESF Logstash plugin](https://rubygems.org/gems/logstash-input-elastic_serverless_forwarder)
`$LOGSTASH_FOLDER/bin/logstash-plugin install logstash-input-elastic_serverless_forwarder`
* Follow [ESF Logstash plugin docs](https://github.com/logstash-plugins/logstash-input-elastic_serverless_forwarder/blob/main/docs/index.asciidoc) to configure the plugin
* Deploy ESF on your AWS account. Refer to the following example configuration
```yaml
inputs:
  - type: "s3-sqs" #or another input type
    id: "arn:aws:sqs:%REGION%:%ACCOUNT%:%QUEUENAME%"
    outputs:
      - type: "logstash"
        args:
          logstash_url: logstash_url
          username: "username" #optional
          password: "password" #optional
          max_batch_size: 500 #optional
          compression_level: 1 #optional
          ssl_assert_fingerprint: "22:F7:FB:84:1D:43:3E:E7:BB:F9:72:F3:D8:97:AD:7C:86:E3:08:42" #optional
```

## Example event as received by Logstash

```
{
    "aws" => {
	    "s3" => {
		    "bucket" => {
			    "name" => "davide-girardi-vpcflowlogs-test",
			    "arn" => "arn:aws:s3:::davide-girardi-vpcflowlogs-test"
		    },
		    "object" => {
			    "key" => "AWSLogs/11111111111/vpcflowlogs/eu-west-1/2023/01/13/1111111111_vpcflowlogs_eu-west-1_fl-0cea9cbf050c152d5_20230113T1610Z_25f82e62.log.gz"
		    }
	    }
    },
    "message" => "2 1111111111 eni-0e854ebb664e54fb7 172.31.37.100 179.60.147.157 22 4954 6 8 480 1673626211 1673626264 ACCEPT OK",
    "cloud" => {
	    "region" => "eu-west-1",
	    "provider" => "aws",
	    "account" => {
		    "id" => "1111111111111111"
	    }
    },
    "log" => {
	    "file" => {
		    "path" => "https://davide-girardi-vpcflowlogs-test.s3.eu-west-1.amazonaws.com/AWSLogs/111111111111/vpcflowlogs/eu-west-1/2023/01/13/11111111111111111_vpcflowlogs_eu-west-1_fl-0cea9cbf050c152d5_20230113T1610Z_25f82e62.log.gz"
	    },
	    "offset" => 2529
    },
    "@timestamp" => 2023 - 01 - 17 T08: 58: 54.995298 Z,
    "@version" => "1",
    "_id" => "timestamp-hash-offset"
}
```

## Related issues

- https://github.com/elastic/obs-infraobs-team/issues/870
